File: workout-tracker/src/app/workouts/models/exercise-plan.ts
import { ResistanceType } from "../enums/resistance-type";
import { SetType } from "../enums/set-type";

export class ExercisePlan {
  public exerciseInWorkoutId: number = -1;
  public exerciseId: number = -1;
  public exerciseName: string = '';

  public numberOfSets: number = 0;
  public setType: SetType = SetType.Repetition;
  public resistanceType: ResistanceType = ResistanceType.BodyWeight;
  public sequence: number = -1;

  public targetRepCountLastTime: number = -1;
  public maxActualRepCountLastTime: number = -1;
  public avgActualRepCountLastTime: number = -1;

  public maxRangeOfMotionLastTime: number = 0;
  public avgRangeOfMotionLastTime: number = 0;

  public maxFormLastTime: number = 0;
  public avgFormLastTime: number = 0;

  public recommendedTargetRepCount: number | null = null;
  public targetRepCount: number | null = null; //Nullable because not all exercises use reps (example: running on a treadmill)

  public resistanceAmountLastTime: number = 0;
  public resistanceMakeupLastTime: string = '';

  public recommendedResistanceAmount: number | null = null;
  public recommendedResistanceMakeup: string | null = null;

  public resistanceAmount: number = 0;
  public resistanceMakeup: string | null = null; //Only applies to resistance band exercises

  public bandsEndToEnd: boolean | null = null;
  public involvesReps: boolean = true;

  public recommendationReason: string = '';
}

File: API/WorkoutTracker.Application/Exercises/Models/ExercisePlan.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WorkoutTracker.Domain.Exercises;

namespace WorkoutTracker.Application.Exercises.Models
{
    /// <summary>
    /// A class used to plan targets (reps and resistance) for an exercise prior to a workout.
    /// This class is not persisted, but rather used to gather information from the user and then 
    /// create the ExecutedWorkout instance.
    /// </summary>
    public class ExercisePlan
    {
        public int ExerciseInWorkoutId { get; set; }
        public int ExerciseId { get; set; }
        public string ExerciseName { get; set; }
        public byte NumberOfSets { get; set; } //TODO: Evaluate. I think this can be removed.
        public SetType SetType { get; set; }
        public ResistanceType ResistanceType { get; set; }
        public byte Sequence { get; set; }

        public byte TargetRepCountLastTime { get; set; }
        public byte MaxActualRepCountLastTime { get; set; }
        public double AvgActualRepCountLastTime { get; set; }

        public byte MaxRangeOfMotionLastTime { get; set; }
        public double AvgRangeOfMotionLastTime { get; set; }

        public byte MaxFormLastTime { get; set; }
        public double AvgFormLastTime { get; set; }

        public byte? RecommendedTargetRepCount { get; set; }
        public byte TargetRepCount { get; set; }

        public decimal ResistanceAmountLastTime { get; set; }
        public string ResistanceMakeupLastTime { get; set; }
        public decimal? RecommendedResistanceAmount { get; set; }
        public string RecommendedResistanceMakeup { get; set; }
        public decimal ResistanceAmount { get; set; }
        public string ResistanceMakeup { get; set; }
        public bool? BandsEndToEnd { get; set; }
        public bool InvolvesReps { get; set; }

        public string RecommendationReason { get; set; }

        public ExercisePlan() { }

        public ExercisePlan(ExerciseInWorkout exercise)
        {
            if (exercise == null)
                throw new ArgumentNullException(nameof(exercise));
            
            ExerciseInWorkoutId = exercise.Id;
            ExerciseId = exercise.Exercise.Id;
            ExerciseName = exercise.Exercise.Name;
            NumberOfSets = exercise.NumberOfSets; //TODO: Evaluate. I think this can be removed.
            SetType = exercise.SetType;
            ResistanceType = exercise.Exercise.ResistanceType;
            BandsEndToEnd = exercise.Exercise.BandsEndToEnd;
            InvolvesReps = exercise.Exercise.InvolvesReps;
            Sequence = exercise.Sequence;
        }

        public void SetLastTimeValues(IEnumerable<ExecutedExercise> executedExercises)
        {
            if (executedExercises == null || !executedExercises.Any())
                return;

            TargetRepCountLastTime = executedExercises.Max(x => x.TargetRepCount);
            MaxActualRepCountLastTime = executedExercises.Max(x => x.ActualRepCount);
            AvgActualRepCountLastTime = Math.Round(executedExercises.Average(x => x.ActualRepCount), 2);

            MaxRangeOfMotionLastTime = executedExercises.Max(x => x.RangeOfMotionRating);
            AvgRangeOfMotionLastTime = Math.Round(executedExercises.Average(x => x.RangeOfMotionRating), 2);

            MaxFormLastTime = executedExercises.Max(x => x.FormRating);
            AvgFormLastTime = Math.Round(executedExercises.Average(x => x.FormRating), 2);

            var executedWithMaxResitance =
                executedExercises.OrderByDescending(x => x.ResistanceAmount).First();

            ResistanceAmountLastTime = executedWithMaxResitance.ResistanceAmount;
            ResistanceMakeupLastTime = executedWithMaxResitance.ResistanceMakeup;
        }

        public void ApplyRecommendation(ExerciseAmountRecommendation recommendation)
        {
            if (recommendation != null)
            {
                RecommendationReason = recommendation.Reason;
                RecommendedResistanceAmount = recommendation.ResistanceAmount;
                RecommendedResistanceMakeup = recommendation.ResistanceMakeup;
                RecommendedTargetRepCount = recommendation.Reps;
            }
        }
    }
}
